### PR TITLE
Total mix stake accounting

### PR DESCRIPTION
**Pull request by durch**
Adds total mix stake value to the contract, and threads it all the way to the `validator-api`. Total mix stake is increased on mixnode bonding and delegation and decreased on unbonding and undelegation. 

- [x] implement total mix stake
- [x] implement total gateway stake
- [x] tests

### CODE REVIEW

File: contracts/mixnet/src/contract.rs
Comment (jstuczyn) [inline contracts/mixnet/src/contract.rs]: I'd perhaps move it explicitly to migrate via `todo!()` macro, so that migration would fail to run without this change (in case somebody misses the comment and wanted to run it)

File: contracts/mixnet/src/queries.rs
Comment (jstuczyn) [inline contracts/mixnet/src/queries.rs]: I'd write a comment here about why the unwrap here is fine - initially it scared me until I looked at the call stack.

File: contracts/mixnet/src/storage.rs
Comment (jstuczyn) [inline contracts/mixnet/src/storage.rs]: If we're doing a checked_add then we shouldn't be really unwraping it, should we? Either deal with failure cases or just change it into `wrapping_add` or `saturating_add`

File: contracts/mixnet/src/storage.rs
Comment (jstuczyn) [inline contracts/mixnet/src/storage.rs]: Same as above

File: validator-api/src/cache/mod.rs
Comment (jstuczyn) [inline validator-api/src/cache/mod.rs:41]: Does it really make sense to wrap it in cache? Also is there no AtomicU128 to get rid of the lock?

**Review by jstuczyn — CHANGES_REQUESTED** [review]
Looks rather good to me - there are only some tiny thing I've mentioned in the code plus the additional two suggestions:

- perhaps introduce the same concept for total gateway stake for consistency and because presumably you will need it for rewards calculations?
- some tests to make sure the global value is always kept in sync in the contract, especially in the presence of the "shadow" delegations. I _think_ it's fine, but I'd feel more confident if it was actually tested

File: validator-api/src/cache/mod.rs
Comment (durch) [reply validator-api/src/cache/mod.rs:41]: Unfortunately no, `AtomicU128` is in a [weird limbo state](https://github.com/rust-lang/rust/issues/32976#issuecomment-893117019)...

File: contracts/mixnet/src/storage.rs
Comment (durch) [reply contracts/mixnet/src/storage.rs]: This is a good point, we likely want saturating add anyway

File: contracts/mixnet/src/contract.rs
Comment (durch) [reply contracts/mixnet/src/contract.rs]: I had no idea `todo!()` was a thing :)

File: contracts/mixnet/src/storage.rs
Comment (durch) [reply contracts/mixnet/src/storage.rs]: Altough saturating has the potential of leading to weirdness, probably better to just handle the error

File: common/client-libs/validator-client/src/client.rs
Comment (jstuczyn) [inline common/client-libs/validator-client/src/client.rs]: I think you should rename it to `gateway_stake` for consistency sake, I don't think we ever refer to gateways by `gt` in any public API

File: common/client-libs/validator-client/src/nymd/mod.rs
Comment (jstuczyn) [inline common/client-libs/validator-client/src/nymd/mod.rs]: Same naming concern

File: common/client-libs/validator-client/src/nymd/mod.rs
Comment (jstuczyn) [inline common/client-libs/validator-client/src/nymd/mod.rs]: And it seems same thing for the `QueryMsg::GetTotalGtStake` if we're renaming things

File: contracts/mixnet/src/contract.rs
Comment (jstuczyn) [inline contracts/mixnet/src/contract.rs:218]: Actually I never asked - is this  going to be in another pull request or something? Because technically after this PR is merged, any other contract changes will be blocked until the migration is implemented

File: contracts/mixnet/src/storage.rs
Comment (jstuczyn) [inline contracts/mixnet/src/storage.rs:73]: Under what conditions can it produce an error? And if it produces an error isn't it technically some undefined behaviour?

File: contracts/mixnet/src/storage.rs
Comment (jstuczyn) [inline contracts/mixnet/src/storage.rs:88]: Sorry to go back to this after we have already resolved it, but upon further thinking the original `unwrap` here is absolutely fine. The reason being is that it's impossible to have higher total mix stake than our entire token supply which is guaranteed to easily fit in an `uint128`

File: validator-api/src/cache/mod.rs
Comment (jstuczyn) [inline validator-api/src/cache/mod.rs]: naming : )

File: validator-api/src/nymd_client.rs
Comment (jstuczyn) [inline validator-api/src/nymd_client.rs]: and bit more of naming

File: contracts/mixnet/src/contract.rs
Comment (durch) [reply contracts/mixnet/src/contract.rs:218]: I think I'll merge this into the tokenomics branch first, then once we're ready to get that done we can preform the migration

File: contracts/mixnet/src/storage.rs
Comment (durch) [reply contracts/mixnet/src/storage.rs:73]: This handles read in case no values have yet been written to storage, primarily for tests, although it will be useful for migration as well

File: contracts/mixnet/src/storage.rs
Comment (durch) [reply contracts/mixnet/src/storage.rs:88]: It resulted in much nicer code, so I think it was a good comment regardless :)

**Pull request merged**
